document incident response (#2589)
sourcegraph/sourcegraph: doc/dev/incidents.md

Examples:

- sourcegraph.com is down or a critical feature is broken (e.g. sign-in, search, code intel).
- A customer reports that their own Sourcegraph instance is down or a critical feature is broken.
- The build on `master` is broken.

## Identifying an incident

Anything that could potentially be an operational incident is, by definition, an operational incident until it is either resolved or verified to not be an operational incident. In other words, assume the worst.

Operational incidents can be reported by anyone (e.g. customers, Sourcegraph teammates) by any means (e.g. Twitter, GitHub, Slack). The first Sourcegraph teammate (regardless of their role) that becomes aware of an incident is responsible for taking a few actions:

1. If the incident was reported by someone outside of Sourcegraph, acknowledge that the incident is being handled.
2. Start an internal communication thread about this incident in the #incidents channel in Slack.
3. Find an engineer to acknowledge ownership of the incident.
    - If you are an engineer and are available/working, or if you are engineering on-call, then you should immediately acknowledge the incident and start working to resolve it.

>     - If you are not an engineer, or if you are not available/working and not on-call, then you should message (in-person, Slack, phone call) available/working engineers until one acknowledges ownership. If you unable to quickly find an owner, default to calling [the engineer who is on-call](https://app.opsgenie.com/schedule/detail/190e2873-1e3b-4350-b67b-2e681d542970).

Reviewer: Typo

```suggestion
    - If you are not an engineer, or if you are not available/working and not on-call, then you should message (in-person, Slack, phone call) available/working engineers until one acknowledges ownership. If you are unable to quickly find an owner, default to calling [the engineer who is on-call](https://app.opsgenie.com/schedule/detail/190e2873-1e3b-4350-b67b-2e681d542970).
```

## Resolving the incident

The goal is to resolve the incident as quickly and safely as possible. Your default action should be to rollback to a known good state instead of trying to identify the exact issue and fixing it.

Here are some useful procedures:

- [Rollback sourcegraph.com](https://github.com/sourcegraph/deploy-sourcegraph-dot-com/blob/release/README.info.md#how-to-rollback-sourcegraphcom)
- [Fix failed database migration on sourcegraph.com](https://github.com/sourcegraph/sourcegraph/tree/master/migrations#dirty-db-schema)
- Revert a broken commit out of master. If a bad commit has already been deployed to sourcegraph.com and is causing problems, rollback the deploy _before_ reverting the commit in master.
    - Revert the commit in a branch and open a PR.
    - Tag the owner of the reverted commit as a reviewer of the PR.
    - Merge the PR as soon as CI passes (don't block on review).

## Learn from the incident

> Document the incident in the [ops log](https://docs.google.com/document/d/1dtrOHs5STJYKvyjigL1kMm6u-W0mlyRSyVxPfKIOfEw/edit) and schedule a [retrospective](retrospectives/index.md).

Reviewer 2: This to me reads as if we should schedule a retrospective for every ops incident, which feels quite heavy to me. I would rather leave that up to the incident owner's discretion. Am I wrong for thinking this?

Author: I updated the wording to make this discretionary.

# Incidents

This document describes how we deal with operational incidents that require time sensitive responses.

Examples:

- sourcegraph.com is down or a critical feature is broken (e.g. sign-in, search, code intel).
- A customer reports that their own Sourcegraph instance is down or a critical feature is broken.
- There is a security issue with Sourcegraph.

> - The build on `master` is broken.

Reviewer 3: ```suggestion
- The `master` build is broken.
```

## Identifying an incident

Anything that could potentially be an operational incident is, by definition, an operational incident until it is either resolved or verified to not be an operational incident. In other words, assume the worst.

Operational incidents can be reported by anyone (e.g. customers, Sourcegraph teammates) by any means (e.g. Twitter, GitHub, Slack). The first Sourcegraph teammate (regardless of their role) that becomes aware of an incident is responsible for taking a few actions:

1. If the incident was reported by someone outside of Sourcegraph, acknowledge that the incident is being handled.
2. Start an internal communication thread about this incident in the #incidents channel in Slack.
3. Find an engineer to acknowledge ownership of the incident.
    - If you are an engineer and are available/working, or if you are engineering on-call, then you should immediately acknowledge the incident and start working to resolve it.

>     - If you are not an engineer, or if you are not available/working and not on-call, then you should message (in-person, Slack, phone call) available/working engineers until one acknowledges ownership. If you are unable to quickly find an owner, default to calling [the engineer who is on-call](https://app.opsgenie.com/schedule/detail/190e2873-1e3b-4350-b67b-2e681d542970).

Reviewer 3: For those that don't have an OpsGenie account, we can use the Slack integration for two things:

1. To know who's on-call: `/genie whoisoncall`
1. To create an alert for the on-call team if one is necessary: `/genie alert THIS IS AN ALERT MSG for ops_team`

## Owning the incident

> If you are owning or asked to own an incident, it is critical that you acknowledge ownership in the #incidents thread in Slack (i.e. "I am on it"), communicate your intended next steps, and post regular updates on your progress.

Reviewer 3: This "owner" role is similar to an [Incident Commander](https://response.pagerduty.com/before/different_roles/#incident-commander-ic). I kind of like that name :-)

Author: eh, I prefer to error on the side of being descriptive instead of having a special name here.

The owner of the incident may delegate tasks to other available/working engineers. This delegated work preempts work unrelated to operational incidents.

> If the issue can not be quickly resolved (via rollback or other means) and if it is a severe problem with sourcegraph.com, then create an issue on sourcegraph/sourcegraph and tweet from the Sourcegraph account (e.g. https://twitter.com/srcgraph/status/1101603205203484672, https://twitter.com/srcgraph/status/1101606401753792512, https://twitter.com/srcgraph/status/1101621105620529153)

Reviewer 3: Additionally, how would you feel about a [status page](https://www.statuspage.io/)?

Author: I don't have anything against status pages like that, but I am hesitant to introduce a new tool before we need it. I think Twitter + GitHub Issues is sufficient for us for now.

Author: I poked around for a bit and think an automated status page would be nice and is something we should think about if/when we adjust our e2e alerts. I added a note here: https://github.com/sourcegraph/sourcegraph/issues/2361#issuecomment-471016470

## Resolving the incident

The goal is to resolve the incident as quickly and safely as possible. Your default action should be to rollback to a known good state instead of trying to identify the exact issue and fixing it.

Here are some useful procedures:

- [Rollback sourcegraph.com](https://github.com/sourcegraph/deploy-sourcegraph-dot-com/blob/release/README.info.md#how-to-rollback-sourcegraphcom)
- [Fix failed database migration on sourcegraph.com](https://github.com/sourcegraph/sourcegraph/tree/master/migrations#dirty-db-schema)
- Revert a broken commit out of master. If a bad commit has already been deployed to sourcegraph.com and is causing problems, rollback the deploy _before_ reverting the commit in master.
    - Revert the commit in a branch and open a PR.
    - Tag the owner of the reverted commit as a reviewer of the PR.
    - Merge the PR as soon as CI passes (don't block on review).

## Learn from the incident

After the incident is resolved:

- Document the incident in the [ops log](https://docs.google.com/document/d/1dtrOHs5STJYKvyjigL1kMm6u-W0mlyRSyVxPfKIOfEw/edit).

> - Create GitHub issues for any appropriate followup work.

Reviewer 3: Depending on the severity, I think a [Post Mortem template](https://github.com/sourcegraph/infrastructure/issues/1039) would be better. We could include a link to it in the Ops Log.

Author: I think the ops log should contain the post mortem and it can link to the followup issues.

    
## Triage

The goal of triage is to either quickly resolve the incident using basic procedures, or quickly identify the right owner.

1. **Acknowledge ownership** of the incident in the relevant Slack thread in the #incidents channel (i.e. "I'm on it").
2. Attempt to resolve the incident by rolling back to a known good state instead of trying to identify and fix the exact issue. **Communicate your intentions in the Slack thread.**
    - [Rollback sourcegraph.com](https://github.com/sourcegraph/deploy-sourcegraph-dot-com/blob/release/README.info.md#how-to-rollback-sourcegraphcom) 
    - Revert a broken commit out of master. If a bad commit has already been deployed to sourcegraph.com and is causing problems, rollback the deploy _before_ reverting the commit in master.
        - Revert the commit in a branch and open a PR.
        - Tag the owner of the reverted commit as a reviewer of the PR.
        - Merge the PR as soon as CI passes (don't block on review).
    - [Fix failed database migration on sourcegraph.com](https://github.com/sourcegraph/sourcegraph/tree/master/migrations#dirty-db-schema)

> 3. If you can not quickly resolve the incident (i.e. <30 minutes), identify the most logical [resolution owner](#ownership) given what you know (this may be yourself) and have that person **acknowledge ownership** in the Slack thread (i.e. "I'm on it").

Author: ```suggestion
3. If rolling back and or reverting commits doesn't resolve the incident, then identify the most logical [resolution owner](#ownership) given what you know (this may be yourself) and have that person **acknowledge ownership** in the Slack thread (i.e. "I'm on it").
```
